thermal.c

```diff
@@ -14,43 +14,333 @@
+		goto err_out;
+	}
+
+	rc = nl_cb_err(cloned_callback, NL_CB_CUSTOM, handle_error, &callback_rc);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: error callback setup failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_cb_set(cloned_callback, NL_CB_ACK, NL_CB_CUSTOM, handle_end, &callback_rc);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: ack callback setup failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_cb_set(cloned_callback, NL_CB_FINISH, NL_CB_CUSTOM, handle_end, &callback_rc);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: finish callback setup failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_cb_set(cloned_callback, NL_CB_VALID, NL_CB_CUSTOM, handle_groupid, &nldata);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: group id callback setup failed.\n");
+		goto err_out;
+	}
+
+	while (callback_rc != 0) {
+		rc = nl_recvmsgs(sock, cloned_callback);
+		if (rc < 0) {
+			log(TO_ALL, LOG_ERR, "thermal: failed to receive messages.\n");
+			goto err_out;
+		}
+	}
 
-	log(TO_ALL, LOG_ERR, "thermal: not yet implemented to establish netlink.\n");
+	group_id = nldata.id;
+	if (group_id < 0) {
+		log(TO_ALL, LOG_ERR, "thermal: invalid group_id was received.\n");
+		goto err_out;
+	}
+
+	rc = nl_socket_add_membership(sock, group_id);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: failed to join the netlink group.\n");
+		goto err_out;
+	}
+
+	goto out;
+
+err_out:
+	error = TRUE;
+out:
+	nl_cb_put(cloned_callback);
+	nlmsg_free(msg);
 	return error;
 }
 
-static gboolean register_netlink_handler(nl_recvmsg_msg_cb_t handler __attribute__((unused)))
+static int handle_thermal_event(struct nl_msg *msg __attribute__((unused)),
+				void *arg __attribute__((unused)))
+{
+	log(TO_ALL, LOG_ERR, "thermal: not yet implemented to process thermal event.\n");
+	return NL_SKIP;
+}
+
+static int handler_for_debug(struct nl_msg *msg __attribute__((unused)),
+			     void *arg __attribute__((unused)))
+{
+	return NL_SKIP;
+}
+
+static gboolean register_netlink_handler(void)
 {
-	gboolean error = TRUE;
+	gboolean error = FALSE;
+	int rc;
+
+	rc = nl_cb_set(callback, NL_CB_SEQ_CHECK, NL_CB_CUSTOM, handler_for_debug, NULL);
+	if (rc) {
+		error = TRUE;
+		log(TO_ALL, LOG_ERR, "thermal: debug handler registration failed.\n");
+		goto out;
+	}
 
-	log(TO_ALL, LOG_ERR, "thermal: not yet implemented to register thermal handler.\n");
+	rc = nl_cb_set(callback, NL_CB_VALID, NL_CB_CUSTOM, handle_thermal_event, NULL);
+	if (rc) {
+		error = TRUE;
+		log(TO_ALL, LOG_ERR, "thermal: thermal handler registration failed.\n");
+	}
+out:
 	return error;
 }
 
+#define INVALID_NL_FD	-1
+
+gboolean receive_thermal_event(gint fd __attribute__((unused)),
+			       GIOCondition condition,
+			       gpointer user_data __attribute__((unused)))
+{
+	gboolean received = FALSE;
+
+	if (condition == G_IO_IN) {
+		int rc;
+
+		rc = nl_recvmsgs(sock, callback);
+		if (!rc)
+			received = TRUE;
+		else
+			log(TO_ALL, LOG_ERR, "thermal: failed to receive thermal event.\n");
```

Comment:
IIRC g_unit_fd_add cancels the subscribed watch if you return FALSE here.  Which is fine, but it means that a failed call to nl_recvmsgs (which will happen if any of your handler fail to parse the message), will cause the entire monitoring system to shut down.  Is that what you want?

Reply:
Yeah, good point. Let me think a little bit more on this.

Reply:
So, if it returns FALSE here, it will remove the source. Even if a message carries NLMSG_ERROR, or its handling fails, the subscription itself should be separately established. There are some side effects of abandoning it but it is okay to keep it I think.

So let me change not to avoid returning FALSE with some comment like this.

Reply:
In the end, I made it returns FALSE after a few failures. I came to consider a possibility to disconnect it. (See the updated code.)